File: DSA/AnshulJain_590014934/Day_86_Question_2.py
def dfs(v, graph, visited, recstack):
    visited[v] = True
    recstack[v] = True
    for u in graph[v]:
        if not visited[u]:
            if dfs(u, graph, visited, recstack):
                return True
        elif recstack[u]:
            return True
    recstack[v] = False
    return False

n = int(input())
m = int(input())
edges = [list(map(int, input().split())) for _ in range(m)]
graph = [[] for _ in range(n)]
for u, v in edges:
    graph[u].append(v)
visited = [False]*n
recstack = [False]*n
cycle = False
for i in range(n):
    if not visited[i]:
        if dfs(i, graph, visited, recstack):
            cycle = True
            break
print("Yes" if cycle else "No")

File: DSA/VaniJain_590012020/Day86Question2.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 100
int adj[MAX][MAX];
int visited[MAX];
int recStack[MAX];
int n; 
int DFS(int node) 
{
    visited[node] = 1;
    recStack[node] = 1;
    for (int i = 0; i < n; i++) 
    {
        if (adj[node][i]) 
        {
            if (!visited[i] && DFS(i))
                return 1;
            else if (recStack[i])
                return 1; 
        }
    }
    recStack[node] = 0; 
    return 0;
}
int main() 
{
    int e;
    printf("Enter number of vertices: ");
    scanf("%d", &n);
    printf("Enter number of edges: ");
    scanf("%d", &e);
    for (int i = 0; i < n; i++) 
    {
        visited[i] = 0;
        recStack[i] = 0;
        for (int j = 0; j < n; j++)
            adj[i][j] = 0;
    }
    printf("Enter edges (u v) format (directed from u to v):\n");
    for (int i = 0; i < e; i++) 
    {
        int u, v;
        scanf("%d %d", &u, &v);
        adj[u][v] = 1;
    }
    int hasCycle = 0;
    for (int i = 0; i < n; i++) 
    {
        if (!visited[i]) 
        {
            if (DFS(i)) 
            {
                hasCycle = 1;
                break;
            }
        }
    if (hasCycle)
        printf("Directed graph contains a cycle.\n");
    else
        printf("Directed graph does not contain a cycle.\n");
    return 0;
}
